src/contracts/wrappers/ERC1155FloorWrapper.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.4;
+
+import {IERC165} from "@0xsequence/erc-1155/contracts/interfaces/IERC165.sol";
+import {IERC1155} from "@0xsequence/erc-1155/contracts/interfaces/IERC1155.sol";
+import {IERC1155TokenReceiver} from "@0xsequence/erc-1155/contracts/interfaces/IERC1155TokenReceiver.sol";
+import {ERC1155} from "@0xsequence/erc-1155/contracts/tokens/ERC1155/ERC1155.sol";
+import {ERC1155MintBurn} from "@0xsequence/erc-1155/contracts/tokens/ERC1155/ERC1155MintBurn.sol";
+import {IERC1155FloorWrapper} from "../interfaces/IERC1155FloorWrapper.sol";
+import {IERC1155Metadata} from "../interfaces/IERC1155Metadata.sol";
+import {IDelegatedERC1155Metadata} from "../interfaces/IDelegatedERC1155Metadata.sol";
+import {WrapperErrors} from "../utils/WrapperErrors.sol";
+
+/**
+ * @notice Allows users to wrap any amount of a ERC-1155 token with a 1:1 ratio.
+ * Therefore each ERC-1155 within a collection can be treated as if fungible.
+ */
+contract ERC1155FloorWrapper is
+    IERC1155FloorWrapper,
+    ERC1155MintBurn,
+    IERC1155Metadata,
+    IERC1155TokenReceiver,
+    WrapperErrors
+{
+    bool private initialized;
+    address internal immutable factory;
+    IERC1155 public token;
+    // This contract only supports a single token id
+    uint256 public constant TOKEN_ID = 0;
+
+    bool private isDepositing;
+
+    constructor() {
+        factory = msg.sender;
+    }
+
+    function initialize(address tokenAddr) external {
+        if (initialized || msg.sender != factory) {
+            revert InvalidInitialization();
+        }
+        token = IERC1155(tokenAddr);
+        initialized = true;
+    }
+
+    modifier onlyDepositing() {
+        if (!isDepositing) {
+            revert InvalidERC1155Received();
+        }
+        delete isDepositing;
+        _;
+    }
+
+    /**
+     * Prevent invalid method calls.
+     */
+    fallback() external {
+        revert UnsupportedMethod();
+    }
+
+    /**
+     * Deposit and wrap ERC-1155 tokens.
+     * @param tokenIds The ERC-1155 token ids to deposit.
+     * @param tokenAmounts The amount of each token to deposit.
+     * @param recipient The recipient of the wrapped tokens.
+     * @param data Data to pass to ERC-1155 receiver.
+     * @notice Users must first approve this contract address on the ERC-1155 contract.
+     */
+    function deposit(uint256[] memory tokenIds, uint256[] memory tokenAmounts, address recipient, bytes calldata data)
```

Comment:
Why not just use the ERC1155 callback for this? With this approach you always need to do an approve, and by using a callback you can also support wrapping by just sending to this contract.

Reply:
This function signature is nice because it allowed deposits to be swapped immediately. 
https://github.com/0xsequence/niftyswap/blob/592efff98fe6135c80c6c7a16113999eb8b60c28/tests_foundry/ERC1155FloorWrapper.test.sol#L198
To enable this with the ERC1155 Received callback I'd need to do some manual parsing of the `data` which feels messy.

Reply:
Keep in mind decoding `data` is a pattern that we already use heavily on Niftyswap, you can implement it in a "clean" way be using `abi.decode` and a struct definition.

Reply:
I was worried about tightly coupling Niftyswap structs to the wrapper implementation. Found a way around it so I prefer this approach now too.